Description: Fix FTBFS with newer subversion
Origin: https://github.com/RapidSVN/RapidSVN/issues/6
Bug-Debian: #796450
Forwarded: not-needed
Applied-Upstream: https://github.com/jsonn/pkgsrc/commit/9a2dbd9a13f94539cebf27d9de18a8014323c87a
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- rapidsvn-0.12.1/src/svncpp/client_ls.cpp.old	2012-06-28 13:39:33.000000000 +0700
+++ rapidsvn-0.12.1/src/svncpp/client_ls.cpp	2016-01-05 07:13:33.309097969 +0700
@@ -29,6 +29,7 @@
 #include "svn_client.h"
 #include "svn_path.h"
 #include "svn_sorts.h"
+#include "svn_version.h"
 //#include "svn_utf.h"
 
 // svncpp
@@ -37,6 +38,8 @@
 #include "svncpp/exception.hpp"
 
 
+#if SVN_VER_MAJOR == 1 && SVN_VER_MINOR < 8
+
 static int
 compare_items_as_paths(const svn_sort__item_t *a, const svn_sort__item_t *b)
 {
@@ -90,8 +93,75 @@
   }
 }
 
+#else
+
+#include <algorithm>
+
+static svn_error_t* store_entry(
+        void *baton,
+        const char *path,
+        const svn_dirent_t *dirent,
+        const svn_lock_t *,
+        const char *abs_path,
+        const char *,
+        const char *,
+        apr_pool_t *scratch_pool)
+{
+  svn::DirEntries *entries = reinterpret_cast<svn::DirEntries*>(baton);
+  if (path[0] == '\0') {
+    if (dirent->kind == svn_node_file) {
+      // for compatibility with svn_client_ls behaviour, listing a file
+      // stores that file name
+      entries->push_back(svn::DirEntry(svn_path_basename(abs_path, scratch_pool), dirent));
+    }
+  } else {
+    entries->push_back(svn::DirEntry(path, dirent));
+  }
+  return SVN_NO_ERROR;
+}
+
+static bool sort_by_path(svn::DirEntry const& a, svn::DirEntry const& b)
+{
+  return svn_path_compare_paths(a.name(), b.name()) < 0;
+}
+
+namespace svn
+{
+  DirEntries
+  Client::list(const char * pathOrUrl,
+               svn_opt_revision_t * revision,
+               bool recurse) throw(ClientException)
+  {
+    Pool pool;
+    DirEntries entries;
+
+    svn_error_t * error =
+      svn_client_list3(pathOrUrl,
+                       revision,
+                       revision,
+                       SVN_DEPTH_INFINITY_OR_IMMEDIATES(recurse),
+                       SVN_DIRENT_ALL,
+                       FALSE, // fetch locks
+                       FALSE, // include externals
+                       &store_entry,
+                       &entries,
+                       *m_context,
+                       pool);
+
+    if (error != SVN_NO_ERROR)
+      throw ClientException(error);
+
+    std::sort(entries.begin(), entries.end(), &sort_by_path);
+
+    return entries;
+  }
+}
+
+#endif
+
 /* -----------------------------------------------------------------
  * local variables:
  * eval: (load-file "../../rapidsvn-dev.el")
  * end:
  */
+
50c50
<     Data(const char * _name, svn_dirent_t * dirEntry)
---
>     Data(const char * _name, const svn_dirent_t * dirEntry)
81c81
<   DirEntry::DirEntry(const char * name, svn_dirent_t * DirEntry)
---
>   DirEntry::DirEntry(const char * name, const svn_dirent_t * DirEntry)
153a154
> 
--- rapidsvn-0.12.1/include/svncpp/dirent.hpp.old	2012-06-28 13:39:20.000000000 +0700
+++ rapidsvn-0.12.1/include/svncpp/dirent.hpp	2016-01-05 07:11:10.337456433 +0700
@@ -41,7 +41,7 @@
     /**
      * constructor for existing @a svn_dirent_t entries
      */
-    DirEntry(const char * name, svn_dirent_t * dirEntry);
+    DirEntry(const char * name, const svn_dirent_t * dirEntry);
 
     /**
      * copy constructor
@@ -91,3 +91,4 @@
  * eval: (load-file "../../rapidsvn-dev.el")
  * end:
  */
+
--- rapidsvn-0.12.1/src/svncpp/dirent.cpp.old	2016-01-05 07:22:32.647100607 +0700
+++ rapidsvn-0.12.1/src/svncpp/dirent.cpp	2016-01-05 07:23:15.778197025 +0700
@@ -47,7 +47,7 @@
     {
     }
 
-    Data(const char * _name, svn_dirent_t * dirEntry)
+    Data(const char * _name, const svn_dirent_t * dirEntry)
         : name(_name), kind(dirEntry->kind), size(dirEntry->size),
         hasProps(dirEntry->has_props != 0),
         createdRev(dirEntry->created_rev), time(dirEntry->time)
@@ -78,7 +78,7 @@
   {
   }
 
-  DirEntry::DirEntry(const char * name, svn_dirent_t * DirEntry)
+  DirEntry::DirEntry(const char * name, const svn_dirent_t * DirEntry)
       : m(new Data(name, DirEntry))
   {
   }
